library/src/G4CMPPhononBoundaryProcess.cc

```diff
@@ -192,128 +227,249 @@ DoReflection(const G4Track& aTrack, const G4Step& aStep,
     G4Track* sec1 = particleChange.GetSecondary(0);
     G4Track* sec2 = particleChange.GetSecondary(1);
 
-    G4ThreeVector vec1 = GetLambertianVector(surfNorm, mode);
-    G4ThreeVector vec2 = GetLambertianVector(surfNorm, mode);
+    G4ThreeVector vec1 = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                                    surfacePoint);
+    G4ThreeVector vec2 = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                                    surfacePoint);
 
     sec1->SetMomentumDirection(vec1);
     sec2->SetMomentumDirection(vec2);
 
     return;
   } else if (random < downconversionProb + specProb) {
-    reflectedKDir = GetReflectedVector(waveVector, surfNorm, mode);
+    reflectedKDir = GetSpecularVector(waveVector, surfNorm, mode, surfacePoint); // Modify surfacePoint & surfNorm in place
     refltype = "specular";
   } else {
-    reflectedKDir = GetLambertianVector(surfNorm, mode);
+    reflectedKDir = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                               surfacePoint);
     refltype = "diffuse";
   }
 
-  G4ThreeVector vdir = theLattice->MapKtoVDir(mode, reflectedKDir);
-  G4double v = theLattice->MapKtoV(mode, reflectedKDir);
+  // Update trackInfo wavevector and particleChange's group velocity and momentum direction
+  // reflectedKDir is in global coordinates here - no conversion needed
+  FillParticleChange(particleChange, aTrack, reflectedKDir);
+  const G4ThreeVector vdir = *particleChange.GetMomentumDirection();
+
+  // If displacement occured: update the particle change and navigator for volume assignment
+  if (refltype == "specular" && *particleChange.GetPosition() != surfacePoint) {
+    FillParticleChange(phParticleChange, aStep, surfacePoint);
+    UpdateNavigatorVolume(aStep, surfacePoint, vdir);
+  }
 
   if (verboseLevel>2) {
-    G4cout << "\n New wavevector direction " << reflectedKDir
-	   << "\n New momentum direction   " << vdir << G4endl;
+    G4cout << "New surface position " << *particleChange.GetPosition()/mm 
+	   << " mm\n New wavevector direction " << reflectedKDir
+	   << "\n New momentum direction " << vdir << G4endl;
   }
 
   // If reflection failed, report problem and kill the track
-  if (!G4CMP::PhononVelocityIsInward(theLattice,mode,reflectedKDir,surfNorm)) {
+  if (!G4CMP::PhononVelocityIsInward(theLattice,mode,reflectedKDir,surfNorm, surfacePoint)) {
+    G4String msg = "Phonon " + refltype + " reflection failed";
+    msg += "\nPhonon mode at time of death: " + *G4PhononPolarization::Label(mode);
```

Comment:
I didn't realize the `Label()` returned a C-style `const char*`, for which there is no operator+().  That's probably why it worked before, because of the grouping of multiple operator+().  Maybe you want to just use the full phonon name and make it one line:
```
  G4String msg = G4PhononPolarization::Name(mode) + " " + refltype + " reflection failed";
```